content/en/blog/_posts/2023-06-29-container-image-signature-verification/index.md

```diff
@@ -0,0 +1,266 @@
+  Normal   BackOff    22s (x2 over 55s)  kubelet            Back-off pulling image "quay.io/crio/unsigned"
+  Warning  Failed     22s (x2 over 55s)  kubelet            Error: ImagePullBackOff
+  Normal   Pulling    9s (x3 over 58s)   kubelet            Pulling image "quay.io/crio/unsigned"
+  Warning  Failed     6s (x3 over 55s)   kubelet            Failed to pull image "quay.io/crio/unsigned": SignatureValidationFailed: Source image rejected: A signature was required, but no signature exists
+  Warning  Failed     6s (x3 over 55s)   kubelet            Error: SignatureValidationFailed
+```
+
+This overall behavior provides a more Kubernetes native experience and does not
+rely on third party software to be installed in the cluster.
+
+There are still a few corner cases to consider: For example, what if you want to
+allow policies per namespace in the same way the policy-controller supports it?
+Well, there is an upcoming CRI-O feature in v1.28 for that! CRI-O will support
+the `--signature-policy-dir` / `signature_policy_dir` option, which defines the
+root path for pod namespace-separated signature policies. This means that CRI-O
+will lookup that path and assemble a policy like `<SIGNATURE_POLICY_DIR>/<NAMESPACE>.json`,
+which will be used on image pull if existing. If no pod namespace is being
+provided on image pull ([via the sandbox config][sandbox-config]), or the
+concatenated path is non-existent, then CRI-O's global policy will be used as
+fallback.
+
+[sandbox-config]: https://github.com/kubernetes/cri-api/blob/e5515a5/pkg/apis/runtime/v1/api.proto#L1448
+
+Another corner case to consider is critical for the correct signature
+verification within container runtimes: The kubelet only invokes container image
+pulls if the image does not already exist on disk. This means, that a
+unrestricted policy from Kubernetes namespace A can allow pulling an image,
+while namespace B is not able to enforce the policy because it already exits on
+the node. Finally, CRI-O has to verify the policy not only on image pull, but
+also on container creation. This fact makes things even a bit more complicated,
+because the CRI does not really pass down the user specified image reference on
+container creation, but more an already resolved iamge ID or digest. A [small
+change to the CRI][pr-118652] can help with that.
+
+[pr-118652]: https://github.com/kubernetes/kubernetes/pull/118652
+
+Now that everything happens within the container runtime, someone has to
+maintain and define the policies to provide a good user experience around that
+feature. The CRDs of the policy-controller are great, while we could imagine that
+a daemon within the cluster can write the policies for CRI-O per namespace. This
+would make any additional hook obsolete and moves the responsibility of
+verifying the image signature to the actual instance which pulls the image. [We
+were evaluating][thread] other possible paths toward a better container image
```

Comment:
```suggestion
evaluated][thread] other possible paths toward a better container image
```

Reply:
Thank you for incorporating the suggestions @saschagrunert! I did another reading of the blog post and I have these additional suggestions that I hope you find useful. Let me know if you have any questions. 

FYI @sftim, @reylejano, @nate-double-u